File: erlang/chapter_2/chapter_2.erl
-module(chapter_2).

-export([bodyRecFib/1, curryTest/0, fib/1, findMin/1,
	 isSorted/1, reverse/1, uncurryTest/2, sumArray/1]).

% Fibonacci sequence

bodyRecFib(N) when N =< 0 -> 0;
bodyRecFib(N) when N == 1 -> 1;
bodyRecFib(N) -> bodyRecFib(N - 1) + bodyRecFib(N - 2).

fib(N) when N =< 0 -> 0;
fib(N) when N == 1 -> 1;
fib(N) -> fib(N, 1, 0, 1).

fib(Goal, Current, _, N) when Goal == Current -> N;
fib(Goal, Current, M, N) ->
    fib(Goal, Current + 1, N, M + N).

% Array operations

sumArray(Array) -> sumArray(Array, 0).

sumArray([], Sum) -> Sum;
sumArray([Head | Tail], Sum) ->
    sumArray(Tail, Head + Sum).

findMin([]) -> 0;
findMin([Head | Tail]) -> findMin(Tail, Head).

findMin([], Min) -> Min;
findMin([Head | Tail], Min) when Head < Min ->
    findMin(Tail, Head);
findMin([_ | Tail], Min) -> findMin(Tail, Min).

reverse(Array) -> reverse(Array, []).

reverse([], Reverse) -> Reverse;
reverse([Head | Tail], Reverse) ->
    reverse(Tail, [Head | Reverse]).

defaultSort(A, B) -> A =< B.

reverseSort(A, B) -> A >= B.

isSorted([Head | Tail]) ->
    isSorted(Head, Tail, fun reverseSort/2).

isSorted(_, [], _) -> true;
isSorted(Previous, [Head | Tail], F) ->
    Complies = F(Previous, Head),
    if Complies -> isSorted(Head, Tail, F);
       true -> false
    end.

% Currying

curry(F) -> fun (A) -> fun (B) -> F(A, B) end end.

isDivisible(A, B) -> A rem B == 0.

% curryTest() -> ((curry(fun isDivisible/2))(12))(3).
curryTest() ->
    F = curry(fun isDivisible/2),
    G = F(12),
    G(3).

uncurryTest(A, B) ->
    F = curry(fun isDivisible/2),
    G = uncurry(F),
    G(A, B).

uncurry(F) ->
    fun (A, B) ->
      (F(A))(B)
    end.

File: erlang/chapter_2/chapter_2.ex
defmodule Chapter2 do

    # Fibonacci sequence

    # Body recursive form
    def bodyRecFib(0), do: 0
    def bodyRecFib(1), do: 1
    def bodyRecFib(n), do: bodyRecFib(n - 1) + bodyRecFib(n - 2)

    # Tail recursive form
    def fib(n) do
        fib(n, 0, 1)
    end
    def fib(n, prev, _cur) when n <= 0 do
        prev
    end
    def fib(n, prev, cur) do
        fib(n - 1, cur, prev + cur)
    end

    # Factorial
    def fac(n), do: fac(n, 1)
    def fac(n, acc) when n <= 0, do: acc
    def fac(n, acc), do: fac(n - 1, n * acc)

    # Tell if an array is sorted
    def isSorted([head | tail], f), do: isSorted(head, tail, f)

    def isSorted(_current, [], _array), do: true
    def isSorted(current, [head | tail], f) do
        if f.(current, head) do
            isSorted(head, tail, f)
        else
            false
        end
    end

    # Currying
    
    def curry(f) do
        fn(a) ->
            fn(b) -> f.(a, b) end
        end
    end

    # Uncurrying

    def uncurry(f) do
        fn(a,b) -> f.(a).(b) end
    end

    # Composing

    def compose(f ,g) do
        fn(a) ->
            f.(g.(a))
        end
    end

end
